test/common/buffer/buffer_fuzz_test.cc

```diff
@@ -0,0 +1,221 @@
+      const uint32_t available = slices[allocated_slices - 1].len_;
+      const uint32_t remainder = shrink_length - target_length;
+      if (available > remainder) {
+        slices[allocated_slices - 1].len_ -= remainder;
+        break;
+      }
+      shrink_length -= available;
+      --allocated_slices;
+    }
+    target_buffer.commit(slices, allocated_slices);
+    ASSERT_EQ(previous_length + target_length, target_buffer.length());
+    break;
+  }
+  case test::common::buffer::Action::kCopyOut: {
+    const uint32_t start = action.copy_out().start() % target_buffer.length();
+    uint8_t copy_buffer[2 * 1024 * 1024];
+    const uint32_t length =
+        std::min(static_cast<uint32_t>(target_buffer.length()),
+                 std::min(action.copy_out().length(), static_cast<uint32_t>(sizeof(copy_buffer))));
+    target_buffer.copyOut(start, length, copy_buffer);
+    const std::string contents = target_buffer.toString();
+    ASSERT_EQ(0, ::memcmp(copy_buffer, contents.data(), length));
+    break;
+  }
+  case test::common::buffer::Action::kDrain: {
+    const uint32_t previous_length = target_buffer.length();
+    const uint32_t drain_length =
+        std::min(static_cast<uint32_t>(target_buffer.length()), action.drain());
+    target_buffer.drain(drain_length);
+    ASSERT_EQ(previous_length - drain_length, target_buffer.length());
+    break;
+  }
+  case test::common::buffer::Action::kLinearize: {
+    target_buffer.linearize(
+        std::min(static_cast<uint32_t>(target_buffer.length()), action.linearize()));
+    break;
+  }
+  case test::common::buffer::Action::kMove: {
+    const uint32_t source_index = action.move().source_index() % BufferCount;
+    if (target_index == source_index) {
+      break;
+    }
+    Buffer::OwnedImpl& source_buffer = buffers[source_index];
+    if (action.move().length() == 0) {
+      target_buffer.move(source_buffer);
+    } else {
+      target_buffer.move(source_buffer, std::min(static_cast<uint32_t>(source_buffer.length()),
+                                                 action.move().length()));
+    }
+    break;
+  }
+  case test::common::buffer::Action::kRead: {
+    int pipe_fds[2] = {0, 0};
+    ASSERT_EQ(0, pipe(pipe_fds));
+    const uint32_t max_length = clampSize(action.read());
+    std::string data(max_length, 'd');
+    const uint32_t previous_length = target_buffer.length();
+    ASSERT_GT(::write(pipe_fds[1], data.data(), max_length), 0);
+    ASSERT_EQ(target_buffer.read(pipe_fds[0], max_length).rc_, max_length);
+    ASSERT_EQ(previous_length, target_buffer.search(data.data(), data.size(), 0));
+    ASSERT_EQ(::close(pipe_fds[0]), 0);
+    ASSERT_EQ(::close(pipe_fds[1]), 0);
+    break;
+  }
+  case test::common::buffer::Action::kWrite: {
+    int pipe_fds[2] = {0, 0};
+    ASSERT_EQ(0, pipe(pipe_fds));
+    const uint32_t previous_length = target_buffer.length();
+    ASSERT_EQ(target_buffer.write(pipe_fds[1]).rc_, previous_length);
```

Comment:
Similarly, if the `previous_length` is really large, this line could block forever.